### PR TITLE
Remove Redundant Host & Device Synchronizations

### LINKED ISSUE #21232: Redundant Host & Device Synchronizations
### Outline & Motivation

We find that PyTorch Lightning can introduce some redundant host & device synchronizations while trying to optimize the performance of some important DL workloads. These syncs affects performance and blocks the usage of CUDA graph.

## Progress Bar Metric

As shown in [logger_connector/result.py#L491-L493](https://github.com/Lightning-AI/pytorch-lightning/blob/2.5.5/src/lightning/pytorch/trainer/connectors/logger_connector/result.py#L491-L493):

```python
# populate progress_bar metrics. convert tensors to numbers
if result_metric.meta.prog_bar:
    metrics["pbar"][forked_name] = convert_tensors_to_scalars(value)
```

If a metric is logged with `prog_bar=True`, e.g. `pl_module.log('lr', lr, prog_bar=True)`, then no matter if user enables progress bar or not, the metric tensor is always converted to scalar, which introduces a host & device sync.

It's better to avoid this conversion if the user don't want to show progress bar, i.e. when `trainer.enable_progress_bar = False`. Then we can avoid such synchronizations.

## Best Metric Device

The metric tensor is always put on the GPU when training with a CUDA device. Thus whenever the metric is retrieved there is a device to host synchronization.

However, I'd propose putting the metric on the device exactly as the value/tensor update it. For example:

* User is likely to update metric `global_step` using a scalar on CPU, so the metric for `global_step` should be on CPU side;
* User is likely to update metric `loss` using a tensor on GPU, so the metric for `loss` should be on GPU side;

In this way, updating `global_step` metric won't introduce a host & device synchronization since the metric is on CPU instead of GPU now. And in future if the user retrieves `global_step` metric, there's also no sync.

The original logic is at [core/module.py#L657-L661](https://github.com/Lightning-AI/pytorch-lightning/blob/2.5.5/src/lightning/pytorch/core/module.py#L657-L661):

```python
value = (
    value.clone().detach()
    if isinstance(value, Tensor)
    else torch.tensor(value, device=self.device, dtype=_get_default_dtype())
)
```

### Pitch

_No response_

### Additional context

_No response_

cc @lantiga @justusschock

### PR DESCRIPTION
**Pull request by alpha0422**
## What does this PR do?

Fixes #21232.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: src/lightning/pytorch/trainer/connectors/logger_connector/result.py [hunk 1]
```diff
@@ -468,7 +468,7 @@ def _forked_name(self, result_metric: _ResultMetric, on_step: bool) -> tuple[str
             forked_name += dataloader_suffix
         return name, forked_name
 
-    def metrics(self, on_step: bool) -> _METRICS:
+    def metrics(self, on_step: bool, *, include_pbar_metrics: bool = True) -> _METRICS:
```
Comment (SkafteNicki) [inline src/lightning/pytorch/trainer/connectors/logger_connector/result.py]:
```suggestion
    def metrics(self, on_step: bool, include_pbar_metrics: bool = True) -> _METRICS:
```
is there a reason the `*` parameter is needed?

**Review by SkafteNicki — CHANGES_REQUESTED** [review]
Can we please add one or more unittests for this?
Mostly to check that we do not regress if changes are made to this part of the codebase?

File: src/lightning/pytorch/trainer/connectors/logger_connector/result.py (hunk 1 above)
Comment (alpha0422) [reply src/lightning/pytorch/trainer/connectors/logger_connector/result.py]: No, we can remove it. Thanks!

**Comment by alpha0422** [comment]
> Can we please add one or more unittests for this? Mostly to check that we do not regress if changes are made to this part of the codebase?

Good point, added a few unittests, thanks.

**Comment by alpha0422** [comment]
Hi @SkafteNicki, what should I do to move forward?

**Comment by SkafteNicki** [comment]
@alpha0422 PR looks fine to me. We are having progrems with our CI at the moment so it may take a bit of time before the PR is merged.

**Comment by anhnguyen-namitech** [comment]
Hope this PR will be merged. We apply this patch, and it speeds up our training a bit.

**Comment by SkafteNicki** [comment]
@anhnami we are doing our best, the hope is #21309 will fix some of the remaining issues that been blocking a lot of PRs the last couple of weeks

**Comment by alpha0422** [comment]
Hi @SkafteNicki, just wanted to follow up on this PR. Is there anything I can do to help move this forward? I'm happy to make any changes if needed.

**Comment by deependujha** [comment]
Hi @alpha0422, can you please take a look at the failing GPU tests!

**Comment by alpha0422** [comment]
Hi @SkafteNicki @deependujha, I have no idea on why some checks fail. It seems one of them failed to comment on the PR, and another three failed without any details. Do you have any guidance on fixing these failures?
